runners/samza/src/main/java/org/apache/beam/runners/samza/renderer/PipelineJsonRenderer.java

```diff
@@ -0,0 +1,186 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.samza.renderer;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Optional;
+import java.util.ServiceLoader;
+import javax.annotation.Nullable;
+import org.apache.beam.model.pipeline.v1.RunnerApi;
+import org.apache.beam.runners.samza.SamzaIOInfo;
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.runners.TransformHierarchy;
+import org.apache.beam.sdk.values.PValue;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.Iterators;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A JSON renderer for BEAM {@link Pipeline} DAG. This can help us with visualization of the Beam
+ * DAG.
+ */
+@Experimental
+public class PipelineJsonRenderer implements Pipeline.PipelineVisitor {
```

Comment:
I suggest to have a inner class in this renderer to implement the visitor instead of letting the rendering implementing it. The methods in the visitor look weird as public methods in the renderer.

Reply:
I've kind of kept this implementation to the way the PipelineDotRenderer is currently implemented right now. Also, doing this would shift most of the code to that inner class, hence I would avoid doing that. What do you think?